src/analyzer.rs

```diff
@@ -9,437 +9,503 @@ use std::collections::{BTreeMap, HashMap};
-                type_: format!("Option<{}>", rust_type),
-                name: key.to_string(),
-                serde_annot: vec![
-                    "default".into(),
-                    "skip_serializing_if = \"Option::is_none\"".into(),
-                ],
-                extra_annot: vec![],
-                docs: member_doc,
-            })
-            // TODO: must capture `default` key here instead of blindly using serde default
-            // this will require us storing default properties for the member in above loop
-            // This is complicated because serde default requires a default fn / impl Default
-            // probably better to do impl Default to avoid having to make custom fns
-        }
+        Ok(dict_key)
     }
-    Ok(Container {
-        name: stack.to_string(),
-        members,
-        level,
-        docs: schema.description.clone(),
-        is_enum: false,
-    })
-}
 
-fn resolve_additional_properties(
-    additional: &JSONSchemaPropsOrBool,
-    stack: &str,
-    key: &str,
-    value: &JSONSchemaProps,
-) -> Result<Option<String>, anyhow::Error> {
-    debug!("got additional: {}", serde_json::to_string(&additional)?);
-    let JSONSchemaPropsOrBool::Schema(s) = additional else {
-        return Ok(None);
-    };
-
-    // This case is for maps. It is generally String -> Something, depending on the type key:
-    let dict_type = s.type_.clone().unwrap_or_default();
-    let dict_key = match dict_type.as_ref() {
-        "string" => Some("String".into()),
-        // We are not 100% sure the array and object subcases here are correct but they pass tests atm.
-        // authoratative, but more detailed sources than crd validation docs below are welcome
-        // https://kubernetes.io/docs/tasks/extend-kubernetes/custom-resources/custom-resource-definitions/#validation
-        "array" => {
-            let mut simple_inner = None;
-            if let Some(JSONSchemaPropsOrArray::Schema(ix)) = &s.items {
-                simple_inner = ix.type_.clone();
-                debug!("additional simple inner  type: {:?}", simple_inner);
-            }
-            // Simple case: additionalProperties contain: {items: {type: K}}
-            // Then it's a simple map (service_monitor_params) - but key is useless
-            match simple_inner.as_deref() {
-                Some("string") => Some("String".into()),
-                Some("integer") => Some(extract_integer_type(s)?),
-                Some("date") => Some(extract_date_type(value)?),
-                Some("") => {
-                    if s.x_kubernetes_int_or_string.is_some() {
-                        Some("IntOrString".into())
-                    } else {
-                        bail!("unknown inner empty dict type for {}", key)
+    // recurse into an array type to find its nested type
+    // this recursion is intialised and ended within a single step of the outer recursion
+    fn array_recurse_for_type(
+        &self,
```

Comment:
perhaps better ad-on suggestion for `Config`: you've indented a bunch of functions here to pass `self` around, but you are ultimately passing the openapi bool around anyway (unless you indent everything, and that doesn't feel much better), maybe it's better to add one extra arg to the top level `analyze_`, a `analyze::Config` and pass that to the few places that need it?

Reply:
yeah i guess it really comes down to a personal preference? all of these functions are inter-related and they all need access to certain configuration, so it made sense to me just make them methods of something that holds the said configuration.

Reply:
yeah, it's fairly equivalent. but it's nice to avoid the right-ward drift